src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -79,10 +80,17 @@ public Dictionary(int capacity, IEqualityComparer<TKey>? comparer)
                 _comparer = comparer;
             }
 
-            if (typeof(TKey) == typeof(string) && _comparer == null)
+            if (typeof(TKey) == typeof(string))
             {
-                // To start, move off default comparer for string which is randomised
-                _comparer = (IEqualityComparer<TKey>)NonRandomizedStringEqualityComparer.Default;
+                if (_comparer == null)
+                {
+                    // To start, move off default comparer for string which is randomised
+                    _comparer = Unsafe.As<IEqualityComparer<TKey>>(NonRandomizedStringEqualityComparer.Default);
+                }
+                else if (_comparer == StringComparer.Ordinal)
```

Comment:
This creates cycle between globalization and Dictionary. It will prevent tree-shaking of StringComparer and its dependencies. @marek-safar Do we have a way to measure the impact of this change on tree-shaking?

Reply:
- This new dependency will cause more stuff to initialize during startup of empty app
- The cycle may lead to subtle bugs (globalization depends on Dictionary, and the Dictionary will depend on globalization). I cannot think of actual bug here, but we had bugs from cycles like these in the past.